huntbugs/src/main/java/one/util/huntbugs/repo/Repository.java

```diff
@@ -59,33 +61,47 @@ public static Repository createSelfRepository() {
         } catch (IOException e) {
             throw new RuntimeException(e);
         }
-        CodeSource codeSource = CompositeRepository.class.getProtectionDomain().getCodeSource();
+
+        Optional<Repository> repo = createDetectorsRepo(CompositeRepository.class, paths);
+        if (repo.isPresent()) {
+            repos.add(repo.get());
+        }
+        return new CompositeRepository(repos);
+    }
+
+    static Repository createPluginRepository(HuntBugsPlugin huntBugsPlugin) {
+        Class<?> pluginClass = huntBugsPlugin.getClass();
+        Optional<Repository> detectorsRepo = createDetectorsRepo(pluginClass, new HashSet<>());
+        return detectorsRepo.isPresent() ? detectorsRepo.get() : createNullRepository();
+    }
+
+    static Optional<Repository> createDetectorsRepo(Class<?> clazz, Set<Path> paths) {
+        CodeSource codeSource = clazz.getProtectionDomain().getCodeSource();
         URL url = codeSource == null ? null : codeSource.getLocation();
         if(url != null) {
             try {
                 Path path = Paths.get(url.toURI());
                 if(paths.add(path)) {
                     if(Files.isDirectory(path))
-                        repos.add(new DirRepository(path));
+                        return Optional.of(new DirRepository(path));
                     else if(Files.isRegularFile(path))
-                        repos.add(new JarRepository(new JarFile(path.toFile())));
+                        return Optional.of(new JarRepository(new JarFile(path.toFile())));
                 }
-            } catch (URISyntaxException | FileSystemNotFoundException | IllegalArgumentException 
+            } catch (URISyntaxException | FileSystemNotFoundException | IllegalArgumentException
                     | IOException | UnsupportedOperationException e) {
                 // ignore
```

Comment:
Isn't it better to propagate exception and fail if something is wrong? I don't see the valid case when empty optional can be removed. It's much better to fail and show an exception to the user than to skip the plugin silently, in which case nobody will understand what's wrong without step-by-step debugging.